wpilibc/src/main/native/cpp/controller/PIDController.cpp

```diff
@@ -0,0 +1,226 @@
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+  m_minimumInput = minimumInput;
+  m_maximumInput = maximumInput;
+  m_inputRange = maximumInput - minimumInput;
+
+  // Clamp setpoint to new input range
+  if (m_maximumInput > m_minimumInput) {
+    m_setpoint = std::clamp(m_setpoint, m_minimumInput, m_maximumInput);
+  }
+}
+
+void PIDController::SetOutputRange(double minimumOutput, double maximumOutput) {
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+  m_minimumOutput = minimumOutput;
+  m_maximumOutput = maximumOutput;
+}
+
+void PIDController::SetAbsoluteTolerance(double tolerance,
+                                         double deltaTolerance) {
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+  m_toleranceType = Tolerance::kAbsolute;
+  m_tolerance = tolerance;
+  m_deltaTolerance = deltaTolerance;
+}
+
+void PIDController::SetPercentTolerance(double tolerance,
+                                        double deltaTolerance) {
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+  m_toleranceType = Tolerance::kPercent;
+  m_tolerance = tolerance;
+  m_deltaTolerance = deltaTolerance;
+}
+
+double PIDController::GetError() const {
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+  return GetContinuousError(m_currError);
+}
+
+/**
+ * Returns the change in error per second of the PIDController.
+ *
+ * @return The change in error per second.
+ */
+double PIDController::GetDeltaError() const {
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+  return (m_currError - m_prevError) / GetPeriod();
+}
+
+double PIDController::Calculate(double measurement) {
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+
+  m_prevError = m_currError;
+  m_currError = GetContinuousError(m_setpoint - measurement);
+
+  if (m_Ki != 0) {
+    m_totalError = std::clamp(m_totalError + m_currError * GetPeriod(),
+                              m_minimumOutput / m_Ki, m_maximumOutput / m_Ki);
+  }
+
+  m_output = std::clamp(m_Kp * m_currError + m_Ki * m_totalError +
+                            m_Kd * (m_currError - m_prevError) / GetPeriod(),
+                        m_minimumOutput, m_maximumOutput);
+
+  return m_output;
+}
+
+double PIDController::Calculate(double measurement, double setpoint) {
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+
+  // Set setpoint to provided value
+  if (m_maximumInput > m_minimumInput) {
+    m_setpoint = std::clamp(setpoint, m_minimumInput, m_maximumInput);
+  } else {
+    m_setpoint = setpoint;
+  }
+
+  m_prevError = m_currError;
+  m_currError = GetContinuousError(m_setpoint - measurement);
+
+  if (m_Ki != 0) {
+    m_totalError = std::clamp(m_totalError + m_currError * GetPeriod(),
+                              m_minimumOutput / m_Ki, m_maximumOutput / m_Ki);
+  }
+
+  m_output = std::clamp(m_Kp * m_currError + m_Ki * m_totalError +
+                            m_Kd * (m_currError - m_prevError) / GetPeriod(),
+                        m_minimumOutput, m_maximumOutput);
```

Comment:
Should we remove the code duplication with another method that does not lock the locks?